C Data Interface Specification (#5442)
apache/arrow: docs/source/format/CProtocol.rst

=========================
The Arrow C data protocol
=========================

*DRAFT*

Data type description - format strings
======================================

TODO:

Structure definitions
=====================

.. code-block:: c

   struct ArrowArray {
     // Type description
     const char* format;
     const char* metadata;

     // Data description
     int64_t length;
     int64_t null_count;
     int64_t offset;
     int64_t n_buffers;
     int64_t n_children;
     const void **buffers;
     struct ArrowArray** children;
     struct ArrowArray* dictionary;

     // Release callback
     void (*release)(struct ArrowArray*);
     // Opaque producer-specific data
     void* private_data;
   };

The ArrowArray structure
------------------------

Members
'''''''

.. c:member:: const char* ArrowArray.format

   Mandatory.  A null-terminated, ASCII-encoded string describing
   the data type.  If the data type is nested, child types are not
   encoded here but in the :c:member:`ArrowArray.children` arrays.

   Consumers MAY decide not to support all data types, but they
   should document this limitation.

.. c:member:: const char* ArrowArray.metadata

>    Optional.  A null-terminated, ASCII-encoded string describing

Reviewer: I think this is inconsistent with flatbuffers which supports UTF-8,

Author: Perhaps. I don't have a strong feeling here. I assumed that full ASCII would be easier to support (at the cost of a limitation for users), but perhaps UTF8 is reasonable in all languages nowadays :-)

   Buffers of children arrays are not included.

.. c:member:: const void** ArrowArray.buffers

   Mandatory.  A C array of pointers to the start of each physical buffer
   backing this array.  Each `void*` pointer is the physical start of
   a contiguous buffer.  There must be :c:member:`ArrowArray.n_buffers` pointers.

   The producer MUST ensure that each contiguous buffer is large enough to
   represent `length + offset` values encoded according to the
   :ref:`Columnar format specification <format_columnar>`.

   The pointer to the null bitmap buffer, if the data type specifies one,
   MAY be NULL only if :c:member:`ArrowArray.null_count` is 0.

   Buffers of children arrays are not included.

.. c:member:: int64_t ArrowArray.n_children

   Mandatory.  The number of children this array has.  The number of children
   is a function of the data type, as described in the
   :ref:`Columnar format specification <format_columnar>`.

.. c:member:: ArrowArray** ArrowArray.children

   Optional.  A C array of pointers to each child array of this array.
   There must be :c:member:`ArrowArray.n_children` pointers.

   MAY be NULL only if :c:member:`ArrowArray.n_children` is 0.

> .. c:member:: ArrowArray* ArrowArray.dictionary

Reviewer: will child arrays have their own release callback?

Reviewer: i see, covered below.

>    Optional.  A pointer to the underlying array of dictionary values.

Reviewer: is the dictionary considered a child from the point of view of releasing?

Author: Yes.

   MUST be present if the format string indicates a dictionary type.
   MUST be NULL otherwise.

> .. c:member:: void (*ArrowArray.release)(struct ArrowArray*)

Reviewer 2: Making this optional seems unnecessary. We could simplify release semantics by simply noting that producers are free to make the release callback a no-op. Then let a released array be signaled with null release callback instead of null format

Author: Perhaps. Though producing a C callback may involve non-trivial boilerplate in some languages, which is why I thought being able to pass null may be convenient.

.. Licensed to the Apache Software Foundation (ASF) under one
.. or more contributor license agreements.  See the NOTICE file
.. distributed with this work for additional information
.. regarding copyright ownership.  The ASF licenses this file
.. to you under the Apache License, Version 2.0 (the
.. "License"); you may not use this file except in compliance
.. with the License.  You may obtain a copy of the License at

..   http://www.apache.org/licenses/LICENSE-2.0

.. Unless required by applicable law or agreed to in writing,
.. software distributed under the License is distributed on an
.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
.. KIND, either express or implied.  See the License for the
.. specific language governing permissions and limitations
.. under the License.

=========================
The Arrow C data protocol
=========================

Rationale
=========

Apache Arrow aims to be a universal in-memory format for the representation

> of tabular ("columnar") data, but some projects may face an undesirable

Reviewer 3: ```suggestion
of tabular ("columnar") data, but some projects may face a difficult
```

Author: Sounds better indeed.

choice between either depending on a fast-evolving dependency such as the
Arrow C++ library, or having to implement adapters for data interchange
(for example by reimplementing the Arrow IPC format, which is non-trivial
and does not allow cross-runtime zero-copy).

The Arrow C data protocol defines a very small, stable set of C definitions

> that can be easily *copied* in any project's source code and used for columnar

Reviewer 3: copied from where? I don't see a *.c file that I would copy.

Author: From the C declarations which are bit later in the spec. Do you think the wording should be clearer?

Reviewer 3: IMO "easily copied" doesn't mean cut-and-paste from code chunks embedded in a prose document. But either way, no, I wouldn't change this sentence here to clarify that you need to copy from this document, just make that clear below if that's the answer.

* Match the Arrow array concepts closely, to avoid the development of
  yet another marshalling layer.
* Avoid the need for one-to-one adaptation layers such as the limited
  JPype-based bridge between Java and Python.
* Allow integration without an explicit dependency (either at compile-time
  or runtime) on the Arrow software project.

Ideally, the Arrow C data protocol can become a low-level *lingua franca*
for sharing columnar data at runtime and establish Arrow as the universal
building block in the columnar processing ecosystem.

Non-goals
---------

* Expose a C API mimicking operations available in higher-level runtimes
  (such as C++, Java...).
* Data sharing between distinct processes.

Comparison with the Arrow IPC format
------------------------------------

Pros of the C data protocol vs. the IPC format:

* No dependency on Flatbuffers.
* No buffer reassembly (data is already exposed in logical Arrow format).
* Zero-copy by design.
* Easy to reimplement from scratch.

Pros of the IPC format vs. the data protocol:

* Works accross processes and machines.
* Allows data storage and persistency.
* Being a streamable format, has room for composing more features (such as

>   integrity checks, compression...).

Reviewer: compression/encoding should probably be accounted for in this spec.  Perhaps its another struct or another member.  Since it doesn't exist yet it seems OK to wait on this.

Author: Do you think that could fit in the `flag` field? Perhaps I should make the flags a `int64_t`?

Reviewer: Yeah, sorry just getting to that int64_t would be better there I think.

Reviewer: thinking about it some more flags are a partial solution I think.  We would still need a way to describe buffer encodings separately (a strawman would be a separate buffer description string using the same format below, one for each buffer).  But this is conjecture, I think we can update the spec once we get agreement on  the encoding proposal (hopefully before the next release) and update the specification at that point.

Author: Ideally the struct remains ABI-stable, so frozen in its definition. So it would have to be added as e.g. a metadata entry, which is perhaps not the most immediate way to access it.

Reviewer: Agreed, I was thinking we would have until 1.0 release to have it finalized.  If we need to make a decision now I think we can go one of two ways:
1. adding a flag below value for "alternative buffer encodings" and then we can add a formal definition to metadata to describe any additional encodings that are approved.  
2.  Add an additional optional `char*` field to describe encodings that is left undefined for the time being.

+------------------------+---------------------------------------------------+------------+
| Format string          | Arrow data type                                   | Notes      |
+========================+===================================================+============+
| ``+l``                 | list                                              |            |
+------------------------+---------------------------------------------------+------------+
| ``+L``                 | large list                                        |            |
+------------------------+---------------------------------------------------+------------+
| ``+w:123``             | fixed-sized list [123 items]                      |            |
+------------------------+---------------------------------------------------+------------+
| ``+s``                 | struct                                            |            |
+------------------------+---------------------------------------------------+------------+
| ``+m``                 | map                                               | \(2)       |
+------------------------+---------------------------------------------------+------------+

> | ``+u:I,J,...``         | union with type ids I,J...                        |            |

Reviewer: we need to distinguish between sparse and dense?

Author: Oh, good catch.

   itself a 2-child struct type of ``(key, value)``.

Examples
--------

* A dictionary-encoded ``decimal128`` array with ``int16`` indices has format
  string ``s``, and its dependent dictionary array has format string ``d``.
* A ``list<uint64>`` array has format string ``+l``, and its single child
  has format string ``L``.
* A ``struct<ints: int32, floats: float32>`` has format string ``+s``; its two
  children have names ``ints`` and ``floats``, and format strings ``i`` and
  ``f`` respectively.
* A ``map<string, float64>`` array has format string ``+m``; its single child
  has name ``entry`` and format string ``+s``; its two grandchildren have names
  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
* A ``union<ints: int32, floats: float32>`` with type ids ``4, 5`` has format
  string ``+u:4,5``; its two children have names ``ints`` and ``floats``,
  and format strings ``i`` and ``f`` respectively.

Structure definitions
=====================

The following free-standing definitions are enough to support the Arrow
C data protocol in your project.  Like the rest of the Arrow project, they

> are available under the Apache License.

Reviewer: do we need to specify Apache 2.0?

Author: Hmm, probably.

.. code-block:: c

   #define ARROW_FLAG_ORDERED 1
   #define ARROW_FLAG_NULLABLE 2

   struct ArrowArray {
     // Type description
     const char* format;
     const char* name;
     const char* metadata;
     int32_t flags;

     // Data description
     int64_t length;
     int64_t null_count;
     int64_t offset;
     int64_t n_buffers;
     int64_t n_children;
     const void** buffers;
     struct ArrowArray** children;
     struct ArrowArray* dictionary;

     // Release callback
     void (*release)(struct ArrowArray*);
     // Opaque producer-specific data
     void* private_data;
   };

The ArrowArray structure
------------------------

.. c:member:: const char* ArrowArray.format

>    Mandatory.  A null-terminated, UTF8-encoded string describing

Reviewer: isn't null termination incompatible with UTF-8?

Author: No. It just forbids using the unicode character 0 :-)

Reviewer: Just reread the spec you are right as always :)

   the data type.  If the data type is nested, child types are not
   encoded here but in the :c:member:`ArrowArray.children` arrays.

   Consumers MAY decide not to support all data types, but they
   should document this limitation.

.. c:member:: const char* ArrowArray.name

   Optional.  A null-terminated, UTF8-encoded string of the field
   or array name.  This is mainly used to reconstruct child fields
   of nested arrays.

   Producers MAY decide not to provide this information, and consumers
   MAY decide to ignore it.  If omitted, MAY be NULL or an empty string.

.. c:member:: const char* ArrowArray.metadata

   Optional.  A null-terminated, UTF8-encoded string describing
   the type's metadata.  If the data type is nested, child types are not
   encoded here but in the :c:member:`ArrowArray.children` arrays.

>    MUST be a JSON-compatible mapping of UTF8 strings to UTF8 strings.

Reviewer: since it UTF8 would it make sense to have a format of:
<key length><key bytes><value length><value bytes>

This would also save encoding/decoding.

Author: Do you mean the lengths would be decimal-encoded? How would it look like exactly?

Reviewer: I would think lengths would be 2s complement int32 in platform endianness.

Author: Hmm, I don't want to grow a custom binary encoding. The advantage of using JSON is that parsers exist for almost every language. Since metadata is optional, I think it's a reasonable compromise. Let's see if others disagree.

   Whitespace MUST only use the 0x20 character. Example::

      {"key1": "base64-encoded value1", "key2": "base64-encode value2"}

   If omitted, MUST be NULL (not an empty string).

   Consumers MAY choose to ignore metadata.  Even then, they SHOULD keep
   the metadata string around so as to propagate its information to their
   own consumers.

.. c:member:: int32_t ArrowArray.flags

   Optional.  A bitfield of flags enriching the type or array description.
   Its value is computed by OR'ing together the flag values.

>    The following flags are available:

Reviewer: i suppose when added this would be one way of communicating encoding.

Author: Hmm, perhaps. We can probably make it required while mentioning that it's gonna be 0 most of the time.

Additionally, the release callback MUST be idempotent, which is commonly
achieved by setting itself to NULL.

Movability
''''''''''

The consumer can *move* the ``ArrowArray`` struct by bitwise copying (or
shallow member-wise copying).  Then it MUST mark the source struct released
(see "released array" above for how to do it) but **without** calling the
release callback.  This ensures that only one live copy of the struct is
active at any given time and that lifetime is correctly communicated to
the producer.

Only the parent array (such as passed between producer and consumer) may be
moved.  It is not recommended to move the children arrays.  It should also
not be necessary, since the :c:member:`ArrowArray.children` and
:c:member:`ArrowArray.dictionary` are typically heap-allocated by the
producer.

.. note::

   For bitwise copying to work correctly, the pointers inside the struct
   (including private_data) MUST not point inside the struct itself.
   Also, external pointers to the struct MUST not be stored by the producer.
   Instead, the producer MUST use the :c:member:`ArrowArray.private_data`
   member so as to remember any necessary bookkeeping information.

> Example use case

Reviewer 3: (Surely out of scope for this protocol draft but) it would be great to have an integration test demonstrating this use case, a blog post with code illustrating the setup, or both.

Author: Yeah, before adopting this spec, I should implement the C++ adaptation layer and if possible a trivial pure Python example.

  yet another marshalling layer.
* Avoid the need for one-to-one adaptation layers such as the limited
  JPype-based bridge between Java and Python.
* Allow integration without an explicit dependency (either at compile-time
  or runtime) on the Arrow software project.

Ideally, the Arrow C data protocol can become a low-level *lingua franca*
for sharing columnar data at runtime and establish Arrow as the universal
building block in the columnar processing ecosystem.

Non-goals
---------

* Expose a C API mimicking operations available in higher-level runtimes
  (such as C++, Java...).
* Data sharing between distinct processes.

Comparison with the Arrow IPC format
------------------------------------

Pros of the C data protocol vs. the IPC format:

* No dependency on Flatbuffers.
* No buffer reassembly (data is already exposed in logical Arrow format).
* Zero-copy by design.
* Easy to reimplement from scratch.

Pros of the IPC format vs. the data protocol:

* Works accross processes and machines.
* Allows data storage and persistency.
* Being a streamable format, has room for composing more features (such as
  integrity checks, compression...).

> Data type description -- format strings

Author: A JSON parser handles decoding the JSON syntax, but you must still _interpret_ the parsed data. I don't think it's significantly easier to go through JSON object key/value pairs, comparing them with allowed keys, cater for missing keys, etc. than to parse the small format that is proposed here.

Metadata is arbitrary key/value by design in Arrow, and it's optional (a consumer is free to ignore it and/or store it undecoded). So JSON is well-suited in that case.

* A dictionary-encoded ``decimal128`` array with ``int16`` indices has format
  string ``s``, and its dependent dictionary array has format string ``d``.
* A ``list<uint64>`` array has format string ``+l``, and its single child
  has format string ``L``.
* A ``struct<ints: int32, floats: float32>`` has format string ``+s``; its two
  children have names ``ints`` and ``floats``, and format strings ``i`` and
  ``f`` respectively.
* A ``map<string, float64>`` array has format string ``+m``; its single child
  has name ``entry`` and format string ``+s``; its two grandchildren have names
  ``key`` and ``value``, and format strings ``u`` and ``g`` respectively.
* A ``sparse_union<ints: int32, floats: float32>`` with type ids ``4, 5``
  has format string ``+us:4,5``; its two children have names ``ints`` and
  ``floats``, and format strings ``i`` and ``f`` respectively.

Structure definitions
=====================

The following free-standing definitions are enough to support the Arrow
C data protocol in your project.  Like the rest of the Arrow project, they
are available under the Apache License 2.0.

.. code-block:: c

   #define ARROW_FLAG_ORDERED 1
   #define ARROW_FLAG_NULLABLE 2

>    struct ArrowArray {

Author: I know, though both cases exists. For example CPython uses the PyCamelCase convention :-)

     // Type description
     const char* format;
     const char* name;
     const char* metadata;

>      int64_t flags;

Author: Even with a version field, there would be constraints (such as not change the struct size). I'm not sure it's a good idea. Let a new version be a separate spec, IMHO.

>      int64_t flags;

Reviewer 4: Version would dictate how to parse format.

   * ``ARROW_FLAG_ORDERED``: for dictionary-encoded arrays, whether the
     ordering of dictionary indices is semantically meaningful.

   If omitted, MUST be 0.

   Consumers MAY choose to ignore some or all of the flags.  Even then,
   they SHOULD keep this value around so as to propagate its information
   to their own consumers.

.. c:member:: int64_t ArrowArray.length

   Mandatory.  The logical length of the array (i.e. its number of items).

.. c:member:: int64_t ArrowArray.null_count

   Mandatory.  The number of null items in the array.  MAY be -1 if not
   yet computed.

.. c:member:: int64_t ArrowArray.offset

   Mandatory.  The logical offset inside the array (i.e. the number of items
   from the physical start of the buffers).  MUST be 0 or positive.

   Producers MAY specify that they will only produce 0-offset arrays to
   ease implementation of consumer code.
   Consumers MAY decide not to support non-0-offset arrays, but they
   should document this limitation.

.. c:member:: int64_t ArrowArray.n_buffers

   Mandatory.  The number of physical buffers backing this array.  The
   number of buffers is a function of the data type, as described in the

>    :ref:`Columnar format specification <format_columnar>`.

Reviewer 5: I am not sure if this problem should be discussed here: is it possible for a null value to hold non-empty space in the data buffer?
There is a similar discussion for the Java code base:
https://www.mail-archive.com/dev@arrow.apache.org/msg14284.html

Author: I don't think it should be discussed here, no. I'm a bit surprised by Jacques' answer in that discussion, though. I would have said "1. a null value always takes no space".

Reviewer 6: @liyafan82 I'm a bit confused by your question here. I don't see how it's related to the List format discussion. We have taken the position uniformly that we do not specify what data is "underneath" a null. So in a ListVector a non-empty element could be "made null" by applying a different validity bitmap. To do otherwise would be inconsistent

Reviewer 5: @pitrou Thanks for your clarification.
@wesm I think that discussion is also related to the layout of ListVector. In particular, whether a null in a list can have some position in the underlying vector is similar to the problem of whether a null value in VarCharVector can hold some space in the data buffer.

   Buffers of children arrays are not included.

.. c:member:: const void** ArrowArray.buffers

   Mandatory.  A C array of pointers to the start of each physical buffer
   backing this array.  Each `void*` pointer is the physical start of
   a contiguous buffer.  There must be :c:member:`ArrowArray.n_buffers` pointers.

   The producer MUST ensure that each contiguous buffer is large enough to
   represent `length + offset` values encoded according to the
   :ref:`Columnar format specification <format_columnar>`.

   The pointer to the null bitmap buffer, if the data type specifies one,

>    MAY be NULL only if :c:member:`ArrowArray.null_count` is 0.

Reviewer 5: Is it possible that the null_count is 0, and the bitmap buffer is non-empty? 
I want to know if the null bitmap buffer is created lazily, eagerly, or implementation dependent?

Author: It is possible. In some situations you don't know that there are zero nulls upfront, and you have a buffer.

Reviewer 5: I see. Thanks.

-----------------

Member allocation
'''''''''''''''''

While the base ArrowArray struct MAY be stack- or heap-allocated by
the consumer (and then a pointer passed around to the producer), any
data pointed to by the struct MUST be allocated by the producer.  This
includes the format and metadata strings, the arrays of buffer and children
pointers, etc.

Therefore, the consumer MUST not try to interfere with the producer's
handling of these members' lifetime.  The only way the consumer influences
data lifetime is by calling the base ArrowArray's release callback.

Released array
''''''''''''''

A released array is indicated by setting :c:member:`ArrowArray.format` to
NULL.  Consumers SHOULD check for a NULL format string and treat it
accordingly (probably by erroring out).  Additionally, a released array
MAY set :c:member:`ArrowArray.release` to NULL.

Release callback semantics -- for consumers
'''''''''''''''''''''''''''''''''''''''''''

Consumers MUST call an array's release callback when they won't be using
it anymore, but they MUST not call any of its child arrays' release callbacks
(including the optional dictionary).  The producer is responsible for releasing
the children.

> Consumers MUST check that the release callback is non-NULL before calling it.

Author: Yeah, we can probably provide a very small library of inline helper functions. They wouldn't be part of the spec, though.

> Consumers MUST check that the release callback is non-NULL before calling it.

Reviewer 2: One of those could be `arrow_noop_release_callback()`

after calling its release callback -- including any associated data such
as its children.

Release callback semantics -- for producers
'''''''''''''''''''''''''''''''''''''''''''

If producers need additional information for lifetime handling (for
example, a C++ producer may want to use ``shared_ptr`` for array and
buffer lifetime), they MUST use the :c:member:`ArrowArray.private_data`
member to locate the required bookkeeping information.

The release callback MUST not assume that the struct will be located
at the same memory location as when it was originally produced.  The consumer
is free to move the struct around (see "Movability").

The release callback MUST walk all children arrays (including the optional
dictionary) and call their own release callbacks.

The release callback MUST free any data area directly owned by the struct
(such as the buffers and children arrays).

The release callback MUST mark the array as released, by setting
:c:member:`ArrowArray.format` to NULL.

Additionally, the release callback MUST be idempotent, which is commonly
achieved by setting itself to NULL.

Movability
''''''''''

> The consumer can *move* the ``ArrowArray`` struct by bitwise copying (or

Author: Right, see reply above about helper functions.

shallow member-wise copying).  Then it MUST mark the source struct released
(see "released array" above for how to do it) but **without** calling the
release callback.  This ensures that only one live copy of the struct is
active at any given time and that lifetime is correctly communicated to
the producer.

Only the parent array (such as passed between producer and consumer) may be

> moved.  It is not recommended to move the children arrays.  It should also

Reviewer 2: I don't think it's a big can of worms: since moving the child would include marking that child as released, `arrow_array_deinit(&moved_child)` should be a noop

> moved.  It is not recommended to move the children arrays.  It should also

Reviewer 2: Instead of the recommendation against moving children, I would add to the spec that valid arrays are guaranteed to have non-released children. This implies that moving a child consumes the parent as well (so the parent should be deinit'd immediately afterwards). This would mean `arrow_array_deinit()` would be the only operation which must check whether child arrays are released.

> moved.  It is not recommended to move the children arrays.  It should also

Reviewer 2: That's why I would say the parent must be released if the child is moved.

> moved.  It is not recommended to move the children arrays.  It should also

Reviewer 4: The problem with move'd children is that ArrowArray structs are likely to be heap allocated (and owned) by the producer. When you move them, you'd have the consumer (the one who moves) allocate new structs on the heap.

This is the can of worm, for various reason, but mostly about the opaque release function.

> moved.  It is not recommended to move the children arrays.  It should also

Reviewer 2: Moving does not require new heap allocations

> moved.  It is not recommended to move the children arrays.  It should also

Reviewer 2: This really seems like something we'll need to be able to do:

- move a column out of a record batch (or equivalently a StructArray)
- move chunks from a chunked array

> moved.  It is not recommended to move the children arrays.  It should also

Author: @bkietz That's a fair point. I'll think about changing the spec on this.
